src/main/kotlin/com/exactpro/th2/codec/EventBatchCollector.kt

```diff
@@ -0,0 +1,289 @@
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package com.exactpro.th2.codec
+
+import com.exactpro.th2.codec.util.getAllMessages
+import com.exactpro.th2.codec.util.toDebugString
+import com.exactpro.th2.common.grpc.Event
+import com.exactpro.th2.common.grpc.EventBatch
+import com.exactpro.th2.common.grpc.EventID
+import com.exactpro.th2.common.grpc.Message
+import com.exactpro.th2.common.grpc.MessageID
+import com.exactpro.th2.common.grpc.RawMessage
+import com.exactpro.th2.common.schema.message.MessageRouter
+import com.google.protobuf.GeneratedMessageV3
+import mu.KotlinLogging
+import java.time.LocalDateTime
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.Executors
+import java.util.concurrent.ScheduledFuture
+import java.util.concurrent.TimeUnit
+import com.exactpro.th2.common.event.bean.Message as MessageEvent
+
+private val logger = KotlinLogging.logger {}
+
+class CollectorTask {
+    @Volatile
+    lateinit var eventBatchBuilder: EventBatch.Builder
+
+    @Volatile
+    lateinit var scheduledFuture: ScheduledFuture<*>
+
+    @Volatile
+    var isSent: Boolean = false
+
+    fun update(
+        event: Event,
+        scheduleCollectorTask: (collectorTask: CollectorTask) -> ScheduledFuture<*>
+    ) {
+        eventBatchBuilder = EventBatch.newBuilder().setParentEventId(event.parentId)
+        scheduledFuture = scheduleCollectorTask(this)
+        isSent = false
+    }
+}
+
+class EventBatchCollector(
+    private val eventBatchRouter: MessageRouter<EventBatch>,
+    private val maxBatchSize: Int,
+    private val timeout: Long,
+    private val numOfEventBatchCollectorWorkers: Int
+) : AutoCloseable {
+    private val collectorTasks = ConcurrentHashMap<EventID, CollectorTask>()
+    private val scheduler = Executors.newScheduledThreadPool(numOfEventBatchCollectorWorkers)
+
+    private lateinit var rootEventID: EventID
+    private lateinit var decodeErrorGroupEventID: EventID
+    private lateinit var encodeErrorGroupEventID: EventID
+
+    private fun putEvent(event: Event) {
+        val collectorTask = collectorTasks.getOrPut(event.parentId) {
+            CollectorTask().apply {
+                update(event, this@EventBatchCollector::scheduleCollectorTask)
+            }
+        }
+
+        synchronized(collectorTask) {
+            if (collectorTask.isSent) {
+                collectorTasks[event.parentId] = collectorTask.apply {
+                    update(event, this@EventBatchCollector::scheduleCollectorTask)
+                }
+            }
+            collectorTask.eventBatchBuilder.addEvents(event)
+            if (collectorTask.eventBatchBuilder.eventsList.size == maxBatchSize) {
+                collectorTask.scheduledFuture.cancel(true)
+                sendEventBatch(collectorTask.eventBatchBuilder.build())
+                collectorTasks.remove(event.parentId)
```

Comment:
call executeCollectorTask(collectorTask)
```suggestion
                executeCollectorTask(collectorTask)
```

Reply:
There will be deadlock